### DISCUSSION TITLE
Implement MLFlow for deployment and research processes for ML model

### DISCUSSION THREAD

**Discussion by vee-huyvunguyen** [nmh4598/mlops_fraud_detection#9] [Ideas]
### Based on the [sample solution][1] from the judges, which use `MLFlow` for both **deployment** and **training** model, I suggest these changes to the CI/CD workflow:
1. Instead of pushing the **model_exported_in_a_file.file** to repo, we push **only code used for training** (E.g.: `training_model_phase1_prob1.py`)
	- Training processes must be monitored by `mlflow`
2. Setup **GitHub Actions**, to train model, **ONLY** when changes detected in the **training code** 
	- When this happens, the model will be trained on the **raw data** stored in **postgres db** (both **original** data, newly updated data requested **from Judges**, and **other** data source)
3. AI Engineer can **login MLFlow** and **manually train the model** on newly updated data
4. Every time a **new model** is trained (either *manually* or through *GitHub Actions* ), the model must be labeled. (labeling method ?)
	- E.g.: `phase1-prob1-model.github.v2.1`, `phase1-prob1-model.manual.v5.3`, `phase2-prob3-model.nightly` 
5. The API will use `mlflow` to load the **Model** (stored in `mlflow` experiments, previous training processes)
	- Should the API **dynamically decides the model** to serve ? E.g.:
		- Choose the latest version `phase1-prob1-model.latest` , instead of statically labeled model `phase1-prob1-model.v5.3` 
		- Based on **Data Distribution** to choose model ??? (possible ?)
		- etc.

[1]: https://github.com/MLOpsVN/mlops-mara-sample-public

**Comment by nmh4598** [comment]
1. Let see, maybe you are right, if the model is easy to train we don't need to save the model 
2. I think it depends on many factors, so i asked Chatgpt:  
- **Continuous integration (CI)**: You can trigger GitHub Actions on every code push or `pull request` to automatically build, test, and validate your code changes. This can ensure that your code is always in a working state and prevent issues from slipping through to production.
- **Continuous delivery (CD)**: Once your code changes have passed the CI stage, you can trigger GitHub Actions to automatically deploy your `code changes` to a `staging environment`for further testing and validation. This can help catch issues that may only appear in a production-like environment.
- **Model training**: Trigger GitHub Actions to orchestrate the training and validation of machine learning models. This can include running `data processing scripts`, `training scripts`, and `model evaluation scripts`.
- **Model deployment**: Trigger GitHub Actions to automatically deploy `machine learning` models to production environments. This can include deploying models to cloud-based services like `AWS Lambda `or `Kubernetes `clusters. ( In this marathon, maybe we don't need to deploy on AWS, i think it is a bonus if we want )
- **Model monitoring**: Trigger GitHub Actions to automatically monitor the performance of deployed `machine learning models` in production environments. This can include running health checks, `logging metrics`, and alerting on anomalous behavior.

3. I need to think about should we use Airflow now or not, if we use Airflow, we can trigger for the new data, but i still don't understand the method `data drift` from sample solution, it doesn't make sense for me. In phase 2, i don't know the judges give us the `data with or without labels`. 
4. Yes, we should label our model, i think we can use artifact of `MLflow`
5. I don't know yet, but i think the latest model version

Look at this issues : [Link][1]

We should use [Pathlib][2] to make sure they will work on all OS

[1]: https://github.com/MLOpsVN/mlops-mara-sample-public/issues/1
[2]: https://docs.python.org/3/library/pathlib.html
